### DISCUSSION TITLE
why is a newly joined peer not receiving recently gossiped messages.

### DISCUSSION THREAD

**Discussion by revoltez** [libp2p/rust-libp2p#3918] [Q&A] [answered]
i have modified the gossip default configuration to ensure that they keep gossiping messages in case one peer encounter a  network partition or a newly joined peer should get at least the messages up to 500 seconds back, since each heartbeat is one second:
```
	let gossipsub_config = gossipsub::GossipsubConfigBuilder::default()
			.history_length(500)
			.history_gossip(500)
			.gossip_lazy(32)
			.duplicate_cache_time(Duration::from_secs(500))
			.build().unwrap();
```
however its not working, any idea why? 
and what is the difference between` history_length() `and `history_gossip() `and `duplicate_cache_time()`?

**Comment by dariusc93** [comment]
Those functions largely deals with the cache, which would reduce or prevent duplicate messages from being sent or received during the time it is in cache. I dont recall gossip intended design allowing sending previous messages sent to new peers that join, but that may fall on you to implement such a sync mechanism, which was discussed a little at https://github.com/libp2p/rust-libp2p/discussions/3717

**Reply by revoltez** [reply]
so the more i make the history length and history gossip large the less old messages get gossiped?

**Reply by thomaseizinger** [reply, marked as answer]
No, if a peer is not part of the mesh at the time the message is sent, it will never receive it. If you want a newly connected peer to "catch-up" with the message being sent, you will need to implement your own sync mechanism by directly connecting to other peers and download the messages.
